warehouse/organizations/models.py

```diff
@@ -0,0 +1,263 @@
+            name="organizations_valid_name",
+        ),
+    )
+
+    __repr__ = make_repr("name")
+
+    name = Column(Text, nullable=False)
+    normalized_name = orm.column_property(func.normalize_pep426_name(name))
+    display_name = Column(Text, nullable=False)
+    orgtype = Column(
+        Enum(OrganizationType, values_callable=lambda x: [e.value for e in x]),
+        nullable=False,
+    )
+    link_url = Column(URLType, nullable=False)
+    description = Column(Text, nullable=False)
+    is_active = Column(Boolean, nullable=False, default=False)
+    is_approved = Column(Boolean)
+    created = Column(
+        DateTime(timezone=False),
+        nullable=False,
+        server_default=sql.func.now(),
+        index=True,
+    )
+    date_approved = Column(
+        DateTime(timezone=False),
+        nullable=True,
+        onupdate=func.now(),
+    )
+
+    # TODO: Determine if cascade applies to any of these relationships
+    users = orm.relationship(
+        User, secondary=OrganizationRole.__table__, backref="organizations"  # type: ignore # noqa
+    )
+    projects = orm.relationship(
+        "Project", secondary=OrganizationProject.__table__, backref="organizations"  # type: ignore # noqa
+    )
+
+    events = orm.relationship(
+        "OrganizationEvent",
+        backref="organizations",
+        cascade="all, delete-orphan",
+        lazy=True,
+    )
+
+    # TODO:
+    #    def __acl__(self):
+
+    def record_event(self, *, tag, ip_address, additional):
+        session = orm.object_session(self)
+        organization = OrganizationEvent(
+            user=self, tag=tag, ip_address=ip_address, additional=additional
+        )
+        session.add(organization)
+        session.flush()
+
+        return organization
+
+
+class OrganizationEvent(db.Model):
+    __tablename__ = "organization_events"
+
+    organization_id = Column(
+        UUID(as_uuid=True),
+        ForeignKey("organizations.id", deferrable=True, initially="DEFERRED"),
+        nullable=False,
+        index=True,
+    )
+    tag = Column(String, nullable=False)
+    time = Column(DateTime, nullable=False, server_default=sql.func.now())
+    ip_address = Column(String, nullable=False)
+    additional = Column(JSONB, nullable=True)
+
+
+class OrganizationNameCatalog(db.Model):
+
+    __tablename__ = "organization_name_catalog"
+    __table_args__ = (
+        Index("organization_name_catalog_name_idx", "name"),
+        Index("organization_name_catalog_organization_id_idx", "organization_id"),
+        UniqueConstraint(
+            "name",
+            "organization_id",
+            name="_organization_name_catalog_name_organization_uc",
+        ),
+    )
+
+    __repr__ = make_repr("name", "organization_id")
+
+    name = Column(Text, nullable=False)
+    organization_id = Column(
+        ForeignKey("organizations.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+
+
+class OrganizationInvitationStatus(enum.Enum):
```

Comment:
same same, use sqlalchemy `Enum`?

Reply:
I initially patterned this after what was previously done with RoleInvitationStatus =>
https://github.com/pypa/warehouse/blob/3b373b452345aa4a289bb334bfa6f73378899516/warehouse/packaging/models.py#L84-L87
https://github.com/pypa/warehouse/blob/3b373b452345aa4a289bb334bfa6f73378899516/warehouse/packaging/models.py#L102-L105
This is also seems consistent with the examples in the SQLAlchemy docs => https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum
```
import enum
class MyEnum(enum.Enum):
    one = 1
    two = 2
    three = 3

t = Table(
    'data', MetaData(),
    Column('value', Enum(MyEnum))
)
```
So I think we're good with this one and the OrganizationType comment above? Would you agree @ewdurbin?

Reply:
I should have done some research up front rather than ask the question, but [SQLAlchemy has support for proper postgres enums](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#enum-types), I think the biggest advantage is that if we utilize that we get actual types in the DB and it becomes possible to inspect them as well as ensure that columns don't contain references to out of date values since when using a naive enum.Enum we get a varchar/int field that isn't validated.